geps/gep-1651.md

```diff
@@ -0,0 +1,154 @@
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an 
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "internal" 
+(service.namespace.svc) and "external" (service.namespace.example.com). 
+Gateways using the same GatewayClass on the cluster, but ensure that the 
+"internal" service is only reachable within the cluster. This would greatly 
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of 
+layer 7 load balancing for our internal services.
+
+## API
+
+We propose extending the `GatewayAddress` type to allow end-users to specify
+the accessibility of the requested address using a property named `Scope`. 
+
+`GatewayAddress.Value` is now an `optional` field. This allows users to signal
+their intent for the implementation to dynamically choose an address based on
+the attributes `GatewayAddress.Scope` and `GatewayAddress.Type`.
+
+### Predefined Scopes
+
+Implementations MAY implement the following supplied scopes and MUST abide by 
+their defined accessibility
+
+Scope | Accessibility
+-|-
+`Public`|The address is accessible over the public internet
+`ClusterLocal`|The address is accessible only within the local cluster
+
+### Vendor prefixed scopes
+
+Implementations can define custom scopes by specifying a vendor prefix followed
+by a slash `/` and a custom scope name ie. `com.ingress.dev/my-scope`.
+
+### Default scope
+The default value of `scope` is implementation specific. It is RECOMMENDED that
+the default `scope` remains consistent for Gateways with the same
+`gatewayClassName`.
+
+
+## Examples
+
+#### 1. Request a GatewayAddress that is only accessible within a cluster
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  listeners:  
+  - protocol: HTTP
+    port: 80
+  addresses:
+  - scope: ClusterLocal
+```
+#### 2. Request a GatewayAddress with a specific scope and address
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  listeners:  
+  - protocol: HTTP
+    port: 80
+  addresses:
+  - scope: ClusterLocal
+    value: 10.0.0.8
+```
+#### 3. Request a GatewayAddress that is publically accessible
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: prod-web
+spec:
+  gatewayClassName: acme-lb
+  listeners:  
+  - protocol: HTTP
+    port: 80
+  addresses:
+  - scope: Public
+```
+
+## Alternatives
+
+### Introducing new GatewayAddress Types
+
+We could introduce additional `AddressTypes` (ie. `ClusterLocalIPAddress`) but
+this would lead to a combinatorial explosion as new dimensions (ie. IPv6) are
+introduced.
```

Comment:
[NodeAddressType](https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/core#NodeAddressType) consists of five well-defined types that support IPv6, IPv4, hostname, etc. This is a core API that we should consider using to shape Gateway address types.

Reply:
closing as a dupe of https://github.com/kubernetes-sigs/gateway-api/pull/1653#discussion_r1144018780